pkg/shp/cmd/follower/follow.go

```diff
@@ -0,0 +1,229 @@
+}
+
+// GetLogLock returns the mutex used for coordinating access to log buffers.
+func (f *Follower) GetLogLock() *sync.Mutex {
+	return &f.logLock
+}
+
+func (f *Follower) complete(params *params.Params) error {
+	clientset, err := params.ClientSet()
+	if err != nil {
+		return err
+	}
+	f.logTail = tail.NewTail(f.ctx, clientset)
+	f.namespace = params.Namespace()
+	if f.shpClientset, err = params.ShipwrightClientSet(); err != nil {
+		return err
+	}
+	f.kclientset, err = params.ClientSet()
+	if err != nil {
+		return err
+	}
+	to, err := params.RequestTimeout()
+	if err != nil {
+		return err
+	}
+	f.pw, err = reactor.NewPodWatcher(f.ctx, to, f.kclientset, params.Namespace())
+	if err != nil {
+		return err
+	}
+
+	f.pw.WithOnPodModifiedFn(f.OnEvent)
+	f.pw.WithTimeoutPodFn(f.OnTimeout)
+	f.pw.WithNoPodEventsYetFn(f.OnNoPodEventsYet)
+	return nil
+}
+
+// Log prints a message
+func (f *Follower) Log(msg string) {
+	// concurrent fmt.Fprintf(r.ioStream.Out...) calls need locking to avoid data races, as we 'write' to the stream
+	f.logLock.Lock()
+	defer f.logLock.Unlock()
+	fmt.Fprint(f.ioStreams.Out, msg)
+}
+
+// tailLogs start tailing logs for each container name in init-containers and containers, if not
+// started already.
+func (f *Follower) tailLogs(pod *corev1.Pod) {
+	containers := append(pod.Spec.InitContainers, pod.Spec.Containers...)
+	for _, container := range containers {
+		if _, exists := f.tailLogsStarted[container.Name]; exists {
+			continue
+		}
+		f.tailLogsStarted[container.Name] = true
+		f.logTail.Start(pod.GetNamespace(), pod.GetName(), container.Name)
+	}
+}
+
+// stop invoke stop on streaming components.
+func (f *Follower) stop() {
+	f.logTail.Stop()
+	f.pw.Stop()
+}
+
+// OnEvent reacts on pod state changes, to start and stop tailing container logs.
+func (f *Follower) OnEvent(pod *corev1.Pod) error {
+	switch pod.Status.Phase {
+	case corev1.PodRunning:
+		if !f.enteredRunningState {
+			f.Log(fmt.Sprintf("Pod %q in %q state, starting up log tail", pod.GetName(), corev1.PodRunning))
+			f.enteredRunningState = true
+			// graceful time to wait for container start
+			time.Sleep(3 * time.Second)
+			// start tailing container logs
+			f.tailLogs(pod)
+		}
+	case corev1.PodFailed:
+		msg := ""
+		br, err := f.shpClientset.ShipwrightV1alpha1().BuildRuns(pod.Namespace).Get(f.ctx, f.buildRunName, metav1.GetOptions{})
+		switch {
+		case err == nil && br.IsCanceled():
+			msg = fmt.Sprintf("BuildRun %q has been canceled.\n", br.Name)
+		case err == nil && br.DeletionTimestamp != nil:
+			msg = fmt.Sprintf("BuildRun %q has been deleted.\n", br.Name)
+		case pod.DeletionTimestamp != nil:
+			msg = fmt.Sprintf("Pod %q has been deleted.\n", pod.GetName())
+		default:
+			podBytes, err2 := json.MarshalIndent(pod, "", "    ")
+			if err2 != nil {
+				msg = fmt.Sprintf("Pod %q has failed!\n", pod.GetName())
+			} else {
+				msg = fmt.Sprintf("Pod %q has failed!\nPod YAML:\n%s\n", pod.GetName(), string(podBytes))
```

Comment:
```suggestion
				msg = fmt.Sprintf("Pod %q has failed!\nPod JSON:\n%s\n", pod.GetName(), string(podBytes))
```

I am not sure if the whole pod object is helpful. Should we maybe just print the pod status ? Actually, I rarely need to look at the pod object when a BuildRun failed. The pod logs are mostly enough. Especially once we make use of the step-defined error reason and message and see those set, we imo should not print any pod information at all. But something we can refine later.

Reply:
Yeah this was a late addition based on my debug specific to our e2es and the BATS framework, where I found access to the entire pod logs was not readily available for one reason or the other.  

And that quickly reminded me of the fact that we've often encountered such situations in support OpenShift Builds, where even with pod logs, we and have to ask for build pod yaml.

So I have admittedly borrowed from that experience as well.

Now, will the step-defined error reaons/msg etc. serve as the better alternative ... perhaps.  Though I would not consider that an absolute certainty.

My opinion, let's initially side on "extra information" during failures, and have to eventually scale back after repeated evidence it is not needed.

But of course if you feel strongly, or if other here want to chime in, let's continue on this.  Otherwise, let's indulge me at least for this PR :-)